services/discourse_translator/yandex.rb

```diff
@@ -0,0 +1,200 @@
+      ne: 'ne',
+      be: 'be',
+      no: 'no',
+      bn: 'bn',
+      pa: 'pa',
+      my: 'my',
+      pap: 'pap',
+      bg: 'bg',
+      fa: 'fa',
+      bs: 'bs',
+      pl: 'pl',
+      cy: 'cy',
+      pt: 'pt',
+      hu: 'hu',
+      ro: 'ro',
+      vi: 'vi',
+      ru: 'ru',
+      ht: 'ht',
+      ceb: 'ceb',
+      gl: 'gl',
+      sr: 'sr',
+      nl: 'nl',
+      si: 'si',
+      mrj: 'mrj',
+      sk: 'sk',
+      el: 'el',
+      sl: 'sl',
+      ka: 'ka',
+      sw: 'sw',
+      gu: 'gu',
+      su: 'su',
+      da: 'da',
+      tg: 'tg',
+      he: 'he',
+      th: 'th',
+      yi: 'yi',
+      tl: 'tl',
+      id: 'id',
+      ta: 'ta',
+      ga: 'ga',
+      tt: 'tt',
+      it: 'it',
+      te: 'te',
+      is: 'is',
+      tr: 'tr',
+      es: 'es',
+      udm: 'udm',
+      kk: 'kk',
+      uz: 'uz',
+      kn: 'kn',
+      uk: 'uk',
+      ca: 'ca',
+      ur: 'ur',
+      ky: 'ky',
+      fi: 'fi',
+      zh: 'zh',
+      fr: 'fr',
+      ko: 'ko',
+      hi: 'hi',
+      xh: 'xh',
+      hr: 'hr',
+      km: 'km',
+      cs: 'cs',
+      lo: 'lo',
+      sv: 'sv',
+      la: 'la',
+      gd: 'gd',
+      lv: 'lv',
+      et: 'et',
+      lt: 'lt',
+      eo: 'eo',
+      lb: 'lb',
+      jv: 'jv',
+      mg: 'mg',
+      ja: 'ja',
+      ms: 'ms',
+    }
+
+    def self.access_token_key
+      "yandex-translator"
+    end
+
+    def self.access_token
+      SiteSetting.translator_yandex_api_key || (raise TranslatorError.new("NotFound: Yandex API Key not set."))
+    end
+
+    def self.detect(post)
+      post.custom_fields[DiscourseTranslator::DETECTED_LANG_CUSTOM_FIELD] ||= begin
+        query = default_query.merge(
+          "text" => post.raw
+        )
+
+        uri = URI(DETECT_URI)
+        uri.query = URI.encode_www_form(query)
+
+        response_body = result(uri.to_s, "", default_headers)
+
+        response_body["lang"]
+      end
+    end
+
+    def self.translate_supported?(source, target)
+      SUPPORTED_LANG.keys.include?(source) && SUPPORTED_LANG.values.include?(target)
+    end
+
+    def self.translate(post)
+      detected_lang = detect(post)
+
+      if !SUPPORTED_LANG.keys.include?(detected_lang.to_sym) &&
+        !SUPPORTED_LANG.values.include?(detected_lang.to_s)
+
+        raise TranslatorError.new(I18n.t('translator.failed'))
+      end
+
+      translated_text = from_custom_fields(post) do
+        query = default_query.merge(
+          "lang" => "#{detected_lang}-#{locale}",
+          "text" => post.cooked
+        )
+
+        uri = URI(TRANSLATE_URI)
+        uri.query = URI.encode_www_form(query)
+
+        response_body = result(uri.to_s, "", default_headers)
+        response_body["text"][0]
+      end
+
+      [detected_lang, translated_text]
+    end
+
+    private
+
+    def self.locale
+      SUPPORTED_LANG[I18n.locale] || (raise I18n.t("translator.not_supported"))
+    end
+
+    def self.post(uri, body, headers = {})
+      Excon.post(uri, body: body, headers: headers)
```

Comment:
Here you're doing a `POST` request, but the specs expect `GET`. Which one is correct?

Reply:
Sorry, the correct one is POST as stated in their documentation: https://tech.yandex.com/translate/doc/dg/reference/detect-docpage/ https://tech.yandex.com/translate/doc/dg/reference/translate-docpage/